packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/graph_based_core.rs

```diff
@@ -208,4 +526,191 @@ impl Core {
+    ) {
+        let component_node = GraphNode::Component(component_idx);
+        let referent_node = GraphNode::Component(referent_idx);
+
+        // The virtual node corresponding to `referent`'s children is the first child of `component`,
+        // so that `referent`'s children will be the first children of `component`.
+        let component_children_virtual_node = self
+            .structure_graph
+            .get_nth_child(&component_node, 0)
+            .unwrap();
+        let referent_children_virtual_node = self
+            .structure_graph
+            .get_nth_child(&referent_node, 0)
+            .unwrap();
+        self.structure_graph.prepend_edge(
+            &component_children_virtual_node,
+            &referent_children_virtual_node,
+        );
+
+        // Attributes from `referent` with the same name as an attribute from `component`
+        // are added as dependency of the attribute from `component`
+        // so that they will be used as a fallback if component doesn't have those attributes.
+
+        // TODO: we could check if attributes of `component` were specified and then omit the dependency.
+        let component_attributes_virtual_node = self
+            .structure_graph
+            .get_nth_child(&component_node, 1)
+            .unwrap();
+        let referent_attributes_virtual_node = self
+            .structure_graph
+            .get_nth_child(&referent_node, 1)
+            .unwrap();
+
+        let referent_attributes = self.components[referent_idx].get_attribute_names();
+
+        for (attr_idx, attr_name) in self.components[component_idx]
+            .get_attribute_names()
+            .into_iter()
+            .enumerate()
+        {
+            if let Some(ref_attr_idx) = referent_attributes.iter().position(|a| *a == attr_name) {
+                // found a matching attribute. Add a link to the attributes
+                let comp_attr = self
+                    .structure_graph
+                    .get_nth_child(&component_attributes_virtual_node, attr_idx)
+                    .unwrap();
+
+                let ref_attr = self
+                    .structure_graph
+                    .get_nth_child(&referent_attributes_virtual_node, ref_attr_idx)
+                    .unwrap();
+
+                self.structure_graph.add_edge(&comp_attr, &ref_attr);
+            }
+        }
+
+        if self.components[component_idx].get_component_type()
+            == self.components[referent_idx].get_component_type()
+        {
+            // Since the component types is the same, we make each prop of `referent` be a dependency
+            // of the prop from `component`.
+            // This dependency indicates that any state props requested by each prop of `component`
+            // should be the corresponding state prop of `referent`.
+            let component_props_virtual_node = self
+                .structure_graph
+                .get_nth_child(&component_node, 2)
```

Comment:
Seeing these numbers makes me think we should add functions to `DirectedGraph` to retrieve these virtual nodes. At a minimum, we should make an `enum`.

Reply:
Added: `get_component_props()`. Didn't actually need the virtual nodes here.